File: src/formatters/json.py
"""JSON formatter with service availability metadata."""

import json
from datetime import datetime
from typing import Dict, List, Optional


def format_json(service: str, metrics: Dict[str, any]) -> str:
    """Format metrics as JSON with subscription.

    Args:
        service: Service name ('claude' or 'codex')
        metrics: Metrics dict

    Returns:
        JSON string
    """
    output = {
        "service": service,
        "timestamp": datetime.now().isoformat(),
        "subscription_type": metrics.get('subscription_type'),
        "metrics": []
    }

    for name, data in metrics.items():
        # Skip subscription_type in iteration
        if name == 'subscription_type':
            continue

        output["metrics"].append({
            "name": name,
            "used_pct": data["used_pct"],
            "remaining_pct": data["remaining_pct"],
            "resets": data["resets"]
        })

    return json.dumps(output, indent=2)


def format_all_json(claude_metrics: Dict[str, any], codex_metrics: Dict[str, any]) -> str:
    """Format combined Claude and Codex metrics as JSON.

    Args:
        claude_metrics: Claude metrics dict
        codex_metrics: Codex metrics dict

    Returns:
        JSON string with both services
    """
    output = {
        "timestamp": datetime.now().isoformat(),
        "services": {
            "claude": {
                "subscription_type": claude_metrics.get('subscription_type'),
                "metrics": []
            },
            "codex": {
                "subscription_type": codex_metrics.get('subscription_type'),
                "metrics": []
            }
        }
    }

    # Add Claude metrics
    for name, data in claude_metrics.items():
        if name == 'subscription_type':
            continue
        output["services"]["claude"]["metrics"].append({
            "name": name,
            "used_pct": data["used_pct"],
            "remaining_pct": data["remaining_pct"],
            "resets": data["resets"]
        })

    # Add Codex metrics
    for name, data in codex_metrics.items():
        if name == 'subscription_type':
            continue
        output["services"]["codex"]["metrics"].append({
            "name": name,
            "used_pct": data["used_pct"],
            "remaining_pct": data["remaining_pct"],
            "resets": data["resets"]
        })

    return json.dumps(output, indent=2)


def format_combined_json(
    claude_metrics: Optional[Dict[str, any]],
    codex_metrics: Optional[Dict[str, any]],
    available_services: List[str]
) -> str:
    """Format combined metrics with service availability metadata.

    Args:
        claude_metrics: Claude metrics dict (None if not collected)
        codex_metrics: Codex metrics dict (None if not collected)
        available_services: List of available service names

    Returns:
        JSON string with availability metadata
    """
    output = {
        "timestamp": datetime.now().isoformat(),
        "available_services": available_services,
        "services": []
    }

    # Add Claude service
    claude_service = {
        "name": "claude",
        "available": "claude" in available_services,
        "subscription_type": claude_metrics.get('subscription_type') if claude_metrics else None,
        "metrics": []
    }

    if claude_metrics:
        for name, data in claude_metrics.items():
            if name == 'subscription_type':
                continue
            claude_service["metrics"].append({
                "name": name,
                "used_pct": data["used_pct"],
                "remaining_pct": data["remaining_pct"],
                "resets": data["resets"]
            })

    output["services"].append(claude_service)

    # Add Codex service
    codex_service = {
        "name": "codex",
        "available": "codex" in available_services,
        "subscription_type": codex_metrics.get('subscription_type') if codex_metrics else None,
        "metrics": []
    }

    if codex_metrics:
        for name, data in codex_metrics.items():
            if name == 'subscription_type':
                continue
            codex_service["metrics"].append({
                "name": name,
                "used_pct": data["used_pct"],
                "remaining_pct": data["remaining_pct"],
                "resets": data["resets"]
            })

    output["services"].append(codex_service)

    return json.dumps(output, indent=2)

File: packages/core/src/formatters/json.ts
/**
 * JSON formatter with service availability metadata.
 * Port of src/formatters/json.py
 */

import { SESSION_WINDOW_HOURS, WEEKLY_WINDOW_HOURS } from "../constants.js";
import type { MetricsDict, ServiceName, ServiceResourceInfo } from "../types.js";
import { calculateTimeProgress, formatTimeRemaining, parseTimeToDatetime } from "../utils/time.js";

const WINDOW_HOURS: Record<string, number> = {
  session: SESSION_WINDOW_HOURS,
  week_all: WEEKLY_WINDOW_HOURS,
  week_sonnet: WEEKLY_WINDOW_HOURS,
  "5h": SESSION_WINDOW_HOURS,
  weekly: WEEKLY_WINDOW_HOURS,
};

function enrichMetric(
  name: string,
  metric: { used_pct: number; remaining_pct: number; resets: string },
): Record<string, unknown> {
  const windowHours = WINDOW_HOURS[name] ?? 168;
  const timeElapsedPct = Math.round(calculateTimeProgress(metric.resets, windowHours));
  const capacityRemaining = timeElapsedPct - metric.used_pct;
  const resetsAt = parseTimeToDatetime(metric.resets);
  return {
    name,
    used_pct: metric.used_pct,
    remaining_pct: metric.remaining_pct,
    time_elapsed_pct: timeElapsedPct,
    capacity_remaining: capacityRemaining,
    resets: metric.resets,
    time_remaining: formatTimeRemaining(new Date(), resetsAt, windowHours),
  };
}

function capacityOnlyMetric(name: string, metric: { used_pct: number; resets: string }): Record<string, unknown> {
  const windowHours = WINDOW_HOURS[name] ?? 168;
  const timeElapsedPct = Math.round(calculateTimeProgress(metric.resets, windowHours));
  return { name, capacity_remaining: timeElapsedPct - metric.used_pct };
}

type ServiceInfoMap = Partial<Record<ServiceName, ServiceResourceInfo>>;

function buildServiceEnvelope(
  serviceName: ServiceName,
  availableServices: string[],
  metrics: MetricsDict | null,
  sources?: Record<string, string>,
  serviceInfo?: ServiceInfoMap,
): Record<string, unknown> {
  const info = serviceInfo?.[serviceName];
  return {
    name: serviceName,
    available: availableServices.includes(serviceName),
    source: info?.source ?? sources?.[serviceName] ?? null,
    stale: info?.stale ?? false,
    error: info?.error ?? null,
    subscription_type: metrics ? ((metrics.subscription_type as string) ?? null) : null,
    metrics: [] as Array<Record<string, unknown>>,
  };
}

/** Format combined metrics with only capacity_remaining per metric */
export function formatCombinedCapacityJson(
  claudeMetrics: MetricsDict | null,
  codexMetrics: MetricsDict | null,
  availableServices: string[],
  sources?: Record<string, string>,
  serviceInfo?: ServiceInfoMap,
  predictions?: Record<string, Record<string, unknown>>,
): string {
  const output: Record<string, unknown> = {
    timestamp: new Date().toISOString(),
    available_services: availableServices,
    services: [] as Array<Record<string, unknown>>,
  };

  const servicesList = output.services as Array<Record<string, unknown>>;

  const claudeService = buildServiceEnvelope("claude", availableServices, claudeMetrics, sources, serviceInfo);
  if (claudeMetrics) {
    for (const [name, data] of Object.entries(claudeMetrics)) {
      if (name === "subscription_type" || typeof data !== "object" || data === null) continue;
      const metric = data as { used_pct: number; resets: string };
      (claudeService.metrics as Array<Record<string, unknown>>).push(capacityOnlyMetric(name, metric));
    }
  }
  if (predictions?.claude) {
    (claudeService as Record<string, unknown>).prediction = predictions.claude;
  }
  servicesList.push(claudeService);

  const codexService = buildServiceEnvelope("codex", availableServices, codexMetrics, sources, serviceInfo);
  if (codexMetrics) {
    for (const [name, data] of Object.entries(codexMetrics)) {
      if (name === "subscription_type" || typeof data !== "object" || data === null) continue;
      const metric = data as { used_pct: number; resets: string };
      (codexService.metrics as Array<Record<string, unknown>>).push(capacityOnlyMetric(name, metric));
    }
  }
  if (predictions?.codex) {
    (codexService as Record<string, unknown>).prediction = predictions.codex;
  }
  servicesList.push(codexService);

  return JSON.stringify(output, null, 2);
}

/** Format single service metrics as JSON string */
export function formatJson(service: string, metrics: MetricsDict): string {
  const output: Record<string, unknown> = {
    service,
    timestamp: new Date().toISOString(),
    subscription_type: (metrics.subscription_type as string) ?? null,
    metrics: [] as Array<Record<string, unknown>>,
  };

  for (const [name, data] of Object.entries(metrics)) {
    if (name === "subscription_type" || typeof data !== "object" || data === null) {
      continue;
    }
    const metric = data as { used_pct: number; remaining_pct: number; resets: string };
    (output.metrics as Array<Record<string, unknown>>).push(enrichMetric(name, metric));
  }

  return JSON.stringify(output, null, 2);
}

/** Format combined Claude and Codex metrics as JSON string */
export function formatAllJson(claudeMetrics: MetricsDict, codexMetrics: MetricsDict): string {
  const output: Record<string, unknown> = {
    timestamp: new Date().toISOString(),
    services: {
      claude: {
        subscription_type: (claudeMetrics.subscription_type as string) ?? null,
        metrics: [] as Array<Record<string, unknown>>,
      },
      codex: {
        subscription_type: (codexMetrics.subscription_type as string) ?? null,
        metrics: [] as Array<Record<string, unknown>>,
      },
    },
  };

  const services = output.services as Record<
    string,
    { subscription_type: string | null; metrics: Array<Record<string, unknown>> }
  >;

  for (const [name, data] of Object.entries(claudeMetrics)) {
    if (name === "subscription_type" || typeof data !== "object" || data === null) continue;
    const metric = data as { used_pct: number; remaining_pct: number; resets: string };
    services.claude.metrics.push(enrichMetric(name, metric));
  }

  for (const [name, data] of Object.entries(codexMetrics)) {
    if (name === "subscription_type" || typeof data !== "object" || data === null) continue;
    const metric = data as { used_pct: number; remaining_pct: number; resets: string };
    services.codex.metrics.push(enrichMetric(name, metric));
  }

  return JSON.stringify(output, null, 2);
}

/** Format combined metrics with service availability metadata */
export function formatCombinedJson(
  claudeMetrics: MetricsDict | null,
  codexMetrics: MetricsDict | null,
  availableServices: string[],
  sources?: Record<string, string>,
  serviceInfo?: ServiceInfoMap,
  predictions?: Record<string, Record<string, unknown>>,
): string {
  const output: Record<string, unknown> = {
    timestamp: new Date().toISOString(),
    available_services: availableServices,
    services: [] as Array<Record<string, unknown>>,
  };

  const servicesList = output.services as Array<Record<string, unknown>>;

  // Claude service
  const claudeService = buildServiceEnvelope("claude", availableServices, claudeMetrics, sources, serviceInfo);
  if (claudeMetrics) {
    for (const [name, data] of Object.entries(claudeMetrics)) {
      if (name === "subscription_type" || typeof data !== "object" || data === null) continue;
      const metric = data as { used_pct: number; remaining_pct: number; resets: string };
      (claudeService.metrics as Array<Record<string, unknown>>).push(enrichMetric(name, metric));
    }
  }
  if (predictions?.claude) {
    (claudeService as Record<string, unknown>).prediction = predictions.claude;
  }
  servicesList.push(claudeService);

  // Codex service
  const codexService = buildServiceEnvelope("codex", availableServices, codexMetrics, sources, serviceInfo);
  if (codexMetrics) {
    for (const [name, data] of Object.entries(codexMetrics)) {
      if (name === "subscription_type" || typeof data !== "object" || data === null) continue;
      const metric = data as { used_pct: number; remaining_pct: number; resets: string };
      (codexService.metrics as Array<Record<string, unknown>>).push(enrichMetric(name, metric));
    }
  }
  if (predictions?.codex) {
    (codexService as Record<string, unknown>).prediction = predictions.codex;
  }
  servicesList.push(codexService);

  return JSON.stringify(output, null, 2);
}
